File: AkkaPingPong.Tests/when_it_gets_a_ping.cs
using AkkaPingPong.AkkaTestBase;
using AkkaPingPong.ASLTestKit.Mocks;
using AkkaPingPong.Common;
using AkkaPingPong.Core.Actors;
using AkkaPingPong.Core.Messages;
using Autofac;
using System;
using Xunit;

namespace AkkaPingPong.Tests
{
    public class when_it_gets_a_ping : TestKitTestBase
    {
        [Fact]
        public void it_should_do_a_pong()
        {
            //Arrange
            //  var container = new ContainerBuilder().Build();
            // mockFactory = new AkkaMockFactory(container, Sys);
            MockFactory.UpdateContainer((builder) =>
            {
                builder.Register<IPingPongService>(b => new FakePingPongService());
            });
            MockFactory.CreateActor<PingPongActor<MockActor>>();

            //Act
            MockFactory.LocateActor(typeof(PingPongActor<>)).Tell(new PingMessage());

            //Assert
            AwaitAssert(() => ExpectMsg<PingMessageCompleted>(), TimeSpan.FromSeconds(5));
        }

        [Fact]
        public void it_should_do_a_pong_integration()
        {
            //Arrange
            //var container = new ContainerBuilder().Build();
            // mockFactory = new AkkaMockFactory(container, Sys);
            MockFactory.UpdateContainer((builder) =>
            {
                builder.Register<IPingPongService>(b => new FakePingPongService());
            });
            MockFactory.CreateActor<PingPongActor<PingCoordinatorActor<PingActor, PingBlockingActor>>>();

            //Act
            for (var i = 0; i < 10; i++)
            {
                System.Threading.Thread.Sleep(1000);
                MockFactory.LocateActor(typeof(PingPongActor<>)).Tell(new PingMessage());
            }

            //Assert
            AwaitAssert(() => ExpectMsg<PongMessage>(), TimeSpan.FromSeconds(20));
        }

        [Fact]
        public void it_should_do_a_pong_unit1()
        {
            // var container = new ContainerBuilder().Build();
            //  mockFactory = new AkkaMockFactory(container, Sys);
            MockFactory.UpdateContainer((builder) =>
            {
                builder.Register<IPingPongService>(b => new FakePingPongService());
            });
            //Arrange
            MockFactory.CreateActor<PingPongActor<MockActor>>();

            //Act
            for (var i = 0; i < 10; i++)
            {
                System.Threading.Thread.Sleep(1000);

                MockFactory.LocateActor(typeof(PingPongActor<>)).Tell(new PingMessage());
            }

            //Assert
            AwaitAssert(() => ExpectMsg<PingMessageCompleted>(), TimeSpan.FromSeconds(20));
        }

        [Fact]
        public void it_should_do_a_pong_unit2()
        {
            // var container = new ContainerBuilder().Build();
            //  mockFactory = new AkkaMockFactory(container, Sys);
            MockFactory.UpdateContainer((builder) =>
            {
                builder.Register<IPingPongService>(b => new FakePingPongService());
            });
            //Arrange
            MockFactory.CreateActor<PingCoordinatorActor<MockActor1, MockActor2>>();

            //Act
            MockFactory.LocateActor(typeof(PingCoordinatorActor<,>)).Tell(new PingMessage());

            //Assert
            AwaitAssert(() => ExpectMsg<SorryImStashing>(), TimeSpan.FromSeconds(20));
        }
    }
}

File: AkkaPingPong.FSTests/when_it_gets_a_ping.fs
namespace AkkaPingPong.FSTests

open Akka.TestKit
open Akka.TestKit.Xunit2
open Akka.TestKit.TestActors
open AkkaPingPong.ActorSystemLib
open AkkaPingPong.Core.Actors
open AkkaPingPong.Core.Messages
open AkkaPingPong.AkkaTestBase
open AkkaPingPong.ASLTestKit
open AkkaPingPong.ASLTestKit.Mocks
open Xunit;
open System
open Autofac

type public when_it_gets_a_ping() =
  inherit TestKitTestBase()

     [<Fact>]
     member public this.``it should do a pong``() =
        //Arrange

        this.MockFactory.CreateActor<PingPongActor<MockActor>>()  |> ignore
        //Act
        new PingMessage() |> this.MockFactory.LocateActor<PingPongActor<_>>().Tell  |> ignore
        //Assert
        let result = this.AwaitAssert  (fun ()->this.ExpectMsg<PingMessageCompleted>() |> ignore),TimeSpan.FromSeconds(5.0)
        ()

     [<Fact>]
     member public this.``it should do a pong integration``() =
        //Arrange

        this.MockFactory.CreateActor<PingPongActor<PingCoordinatorActor<PingActor, PingBlockingActor>>>() |> ignore
        //Act
        [1..10] |> List.iter (fun i->
                               System.Threading.Thread.Sleep(1000) |> ignore
                               new PingMessage() |> this.MockFactory.LocateActor<PingPongActor<_>>().Tell
                               )
        //Assert
        let result = this.AwaitAssert  (fun ()->this.ExpectMsg<PongMessage>() |> ignore),TimeSpan.FromSeconds(20.0)
        ()

     [<Fact>]
     member public this.``it should do a pong unit1``() =
        //Arrange

        this.MockFactory.CreateActor<PingPongActor<MockActor>>() |> ignore
        //Act
        [1..10] |> List.iter (fun i->
                               System.Threading.Thread.Sleep(1000) |> ignore
                               new PingMessage() |> this.MockFactory.LocateActor<PingPongActor<_>>().Tell
                               )
        //Assert
        let result = this.AwaitAssert  (fun m->this.ExpectMsg<PingMessageCompleted>() |> ignore),TimeSpan.FromSeconds(20.0)
        ()

     [<Fact>]
     member public this.``it should do a pong unit2``() =
        //Arrange

        this.MockFactory.CreateActor<PingCoordinatorActor<MockActor1, MockActor2>>()  |> ignore
        //Act
        new PingMessage() |> this.MockFactory.LocateActor<PingCoordinatorActor<_,_>>().Tell  |> ignore
        //Assert
        let result = this.AwaitAssert  (fun ()->this.ExpectMsg<SorryImStashing>() |> ignore),TimeSpan.FromSeconds(5.0)
        ()